File: src/ai/encoding.ts
import { CHECKERS_PER_PLAYER, opponent, POINT_COUNT } from '../engine/helpers';
import { BoardState, Player, PointState } from '../engine/types';

/**
 * TD-Gammon-style board features, always encoded from the perspective of
 * `hero`, the player about to roll — as if hero were white (hero travels
 * toward feature slot 1, hero's home is slots 1–6). The net only ever
 * learns one orientation; black positions are read through the mirror.
 *
 * Layout:
 *   [0..191]  24 slots × (4 hero + 4 opponent) units. For a stack of n:
 *             [n≥1, n≥2, n≥3, max(0, n−3)/2]
 *   [192]     hero checkers on the bar / 2
 *   [193]     opponent checkers on the bar / 2
 *   [194]     hero checkers borne off / 15
 *   [195]     opponent checkers borne off / 15
 *
 * No side-to-move feature: hero is by definition on roll.
 */
export const NUM_INPUTS = 196;

export const ENCODING_ID = 'v0-196';

export function encodeBoard(
  board: BoardState,
  hero: Player,
  out?: Float32Array,
): Float32Array {
  const x = out ?? new Float32Array(NUM_INPUTS);
  if (out) x.fill(0);
  const villain = opponent(hero);
  for (let slot = 1; slot <= POINT_COUNT; slot++) {
    const physical = hero === 'white' ? slot : POINT_COUNT + 1 - slot;
    const point = board.points[physical - 1];
    if (!point) continue;
    const base = (slot - 1) * 8 + (point.player === hero ? 0 : 4);
    const n = point.count;
    x[base] = 1;
    if (n >= 2) x[base + 1] = 1;
    if (n >= 3) x[base + 2] = 1;
    if (n > 3) x[base + 3] = (n - 3) / 2;
  }
  x[192] = board.bar[hero] / 2;
  x[193] = board.bar[villain] / 2;
  x[194] = board.off[hero] / CHECKERS_PER_PLAYER;
  x[195] = board.off[villain] / CHECKERS_PER_PLAYER;
  return x;
}

/**
 * The same position with colors and direction swapped (point p ↔ 25−p).
 * By construction encodeBoard(b, p) === encodeBoard(mirrorBoard(b), opponent(p)).
 */
export function mirrorBoard(board: BoardState): BoardState {
  const points: PointState[] = Array.from({ length: POINT_COUNT }, () => null);
  board.points.forEach((point, i) => {
    points[POINT_COUNT - 1 - i] = point
      ? { player: opponent(point.player), count: point.count }
      : null;
  });
  return {
    points,
    bar: { white: board.bar.black, black: board.bar.white },
    off: { white: board.off.black, black: board.off.white },
  };
}

File: python/bgtrain/encoding.py
"""Batched feature encoder — an exact match of `src/ai/encoding.ts`.

Produces 196 inputs per position from the on-roll player's (hero's)
perspective: the board is read as if hero were white (hero's home is slots
1..6), so the net only ever learns one orientation.
"""

from __future__ import annotations

import torch

from .engine import BAR_B, BAR_W, CHECKERS, OFF_B, OFF_W, POINTS

NUM_INPUTS = 196
ENCODING_ID = "v0-196"


def encode(boards: torch.Tensor, players: torch.Tensor) -> torch.Tensor:
    """``boards`` (N,28) int, ``players`` (N,) sign. Returns (N,196) float32."""
    N = boards.shape[0]
    white = (players > 0).view(N, 1)
    pts = boards[:, :POINTS]  # (N,24) signed, +white

    # Hero-oriented points: positive = hero, in hero-slot order. White reads
    # points as-is; black reverses (slot j -> physical point 25-j) and negates.
    hpts = torch.where(white, pts, -pts.flip(1))  # (N,24)
    hero = hpts.clamp(min=0)  # (N,24)
    opp = (-hpts).clamp(min=0)

    def units(c: torch.Tensor) -> torch.Tensor:  # (N,24) -> (N,24,4)
        return torch.stack(
            [
                (c >= 1).float(),
                (c >= 2).float(),
                (c >= 3).float(),
                (c - 3).clamp(min=0).float() / 2.0,
            ],
            dim=-1,
        )

    slot_feats = torch.cat([units(hero), units(opp)], dim=-1)  # (N,24,8)
    x = torch.empty(N, NUM_INPUTS, dtype=torch.float32, device=boards.device)
    x[:, :192] = slot_feats.reshape(N, 192)

    bar_hero = torch.where(white.view(N), boards[:, BAR_W], boards[:, BAR_B]).float()
    bar_opp = torch.where(white.view(N), boards[:, BAR_B], boards[:, BAR_W]).float()
    off_hero = torch.where(white.view(N), boards[:, OFF_W], boards[:, OFF_B]).float()
    off_opp = torch.where(white.view(N), boards[:, OFF_B], boards[:, OFF_W]).float()
    x[:, 192] = bar_hero / 2.0
    x[:, 193] = bar_opp / 2.0
    x[:, 194] = off_hero / CHECKERS
    x[:, 195] = off_opp / CHECKERS
    return x
